通过回调管理异步和并发主要缺陷：缺乏顺序性和可信任性。解决——promise。

## 什么是Promise
回调如何处理未来值，例子：X+Y？

    function add(getX,getY,cb) {
    var x, y;
    getX( function(xVal){
    x = xVal;
    // 两个都准备好了？
    if (y != undefined) {
    cb( x + y ); // 发送和
    }
    } );
    getY( function(yVal){
    y = yVal;
    // 两个都准备好了？
    if (x != undefined) {
    cb( x + y ); // 发送和
    }
    } );
    }
    // fetchX() 和fetchY()是同步或者异步函数
    add( fetchX, fetchY, function(sum){
    console.log( sum ); // 是不是很容易？
    } ); 

在这段代码中，我们把 x 和 y 当作未来值，并且表达了一个运算 add(..)。这个运算（从外部看）不在意 x 和 y 现在是否都已经可用。换句话说，它把现在和将来归一化了，因此我们可以确保这个 add(..) 运算的输出是可预测的。为了统一处理现在和将来，我们把它们都变成了将来，即所有的操作都成了异步的。

Promise是如何处理未来值，例子：X+Y？

    function add(xPromise,yPromise) {
    // Promise.all([ .. ])接受一个promise数组并返回一个新的promise，
    // 这个新promise等待数组中的所有promise完成
    return Promise.all( [xPromise, yPromise] )
    // 这个promise决议之后，我们取得收到的X和Y值并加在一起
    .then( function(values){
    // values是来自于之前决议的promisei的消息数组
    return values[0] + values[1];
    } );
    }
    // fetchX()和fetchY()返回相应值的promise，可能已经就绪，
    // 也可能以后就绪
    add( fetchX(), fetchY() )
    // 我们得到一个这两个数组的和的promise
    // 现在链式调用 then(..)来等待返回promise的决议
    .then( function(sum){
    console.log( sum ); // 这更简单！
    } ); 

fetchX() 和 fetchY() 是直接调用的，它们的返回值（promise ！）被传给 add(..)。这些promise 代表的底层值的可用时间可能是现在或将来，但不管怎样，promise 归一保证了行为的一致性。我们可以按照不依赖于时间的方式追踪值 X 和 Y。它们是未来值。

add(..)（通过 Promise.all([ .. ])）创建并返回的 promise。我们通过调用then(..) 等待这个 promise。add(..) 运算完成后，未来值 sum 就准备好了，可以打印出来。

在 add(..) 内部，Promise.all([ .. ]) 调用创建了一个 promise（这个promise 等待 promiseX 和 promiseY 的决议）。链式调用 .then(..) 创建了另外一个 promise。这个 promise 由 return values[0] + values[1] 这一行立即决议（得到加运算的结果）。因此，链 add(..) 调用终止处的调用then(..)——在代码结尾处——实际上操作的是返回的第二个 promise，而不是由 Promise.all([ .. ]) 创建的第一个 promise。还有，尽管第二个then(..) 后面没有链接任何东西，但它实际上也创建了一个新的 promise，如果想要观察或者使用它的话就可以看到。

Promise 是一种封装和组合未来值的易于复用的机制。从另外一个角度看待Promise 的决议：一种在异步任务中作为两个或更多步骤的流程控制机制，时序上的 thisthen-that。

    function foo(x) {
    // 开始做点可能耗时的工作
    // 构造一个listener事件通知处理对象来返回
    return listener;
    }
    var evt = foo( 42 );
    evt.on( "completion", function(){
    // 可以进行下一步了！
    } );
    evt.on( "failure", function(err){
    // 啊，foo(..)中出错了
    } );

相对于面向回调的代码，这里的反转是显而易见的，而且这也是有意为之。这里没有把回调传给 foo(..)，而是返回一个名为 evt 的事件注册对象，由它来接受回调。从本质上说，evt 对象就是分离的关注点之间一个中立的第三方协商机制。

resolve 和 reject是 promise 的决议函数。resolve(..) 通常标识完成，而 reject(..) 则标识拒绝。

## Promise 信任问题
Promise 模式构建的可能最重要的特性：信任。

把一个回调传入工具 foo(..) 时可能出现如下问题：

    • 调用回调过早；
    • 调用回调过晚（或不被调用）；
    • 调用回调次数过少或过多；
    • 未能传递所需的环境和参数；
    • 吞掉可能出现的错误和异常。

Promise 的特性就是专门用来为这些问题提供一个有效的可复用的答案。

对于用 Promise.resolve(..) 为所有函数的返回值（不管是不是 thenable）都封装一层。另一个好处是，这样做很容易把函数调用规范为定义良好的异步任务。如果 foo(42) 有时会返回一个立即值，有时会返回 Promise，那么Promise.resolve( foo(42) ) 就能够保证总会返回一个 Promise 结果。而且避免 Zalgo 就能得到更好的代码。

我们可以把多个 Promise 连接到一起以表示一系列异步步骤。

这种方式可以实现的关键在于以下两个 Promise 固有行为特性：

    • 每次你对 Promise 调用 then(..)，它都会创建并返回一个新的 Promise，我们可以将其链接起来；
    • 不管从 then(..) 调用的完成回调（第一个参数）返回的值是什么，它都会被自动设置为被链接 Promise（第一点中的）的完成。

让我们来简单总结一下使链式流程控制可行的 Promise 固有特性。

    • 调用 Promise 的 then(..) 会自动创建一个新的 Promise 从调用返回。
    • 在完成或拒绝处理函数内部，如果返回一个值或抛出一个异常，新返回的（可链接的）
    Promise 就相应地决议。
    • 如果完成或拒绝处理函数返回一个 Promise，它将会被展开，这样一来，不管它的决议值是什么，都会成为当前 then(..) 返回的链接 Promise 的决议值。

术语决议（resolve）、完成（fulfill）和拒绝（reject）。

Promise.resolve(..) 是一个精确的好名字，因为它实际上的结果可能是完成或拒绝。对 then(..) 的第一个参数来说，毫无疑义，总是处理完成的情况，所以不需要使用标识两种状态的术语“resolve”。这里提一下，ES6 规范将这两个回调命名为 onFulfilled(..) 和onRjected(..)，所以这两个术语很准确。

错误处理try..catch 结构，它只能是同步的，无法用于异步代码模式。

为了避免丢失被忽略和抛弃的 Promise 错误，一些开发者表示，Promise 链的一个最佳实践就是最后总以一个 catch(..) 结束，比如：

    var p = Promise.resolve( 42 );
    p.then(
    function fulfilled(msg){
    // 数字没有string函数，所以会抛出错误
    console.log( msg.toLowerCase() );
    }
    )
    .catch( handleErrors ); 

## Promise 模式
在异步序列中（Promise 链），任意时刻都只能有一个异步任务正在执行——步骤 2 只能在步骤 1 之后，步骤 3 只能在步骤 2 之后。如果想要同时执行两个或更多步骤（也就是“并行执行”），要怎么实现呢？Promise.all([ .. ])。

尽管 Promise.all([ .. ]) 协调多个并发 Promise 的运行，并假定所有 Promise 都需要完成，但有时候你会想只响应“第一个跨过终点线的 Promise”，而抛弃其他 Promise。Promise.race([ .. ])。要注意，永远不要递送空数组。

## Promise API 概述
1. new Promise(..) 构造器

有启示性的构造器 Promise(..) 必须和 new 一起使用，并且必须提供一个函数回调。这个回调是同步的或立即调用的。这个函数接受两个函数回调，用以支持 promise 的决议。通常我们把这两个函数称为 resolve(..) 和 reject(..)：

    var p = new Promise( function(resolve,reject){
    // resolve(..)用于决议/完成这个promise
    // reject(..)用于拒绝这个promise
    } );

reject(..) 就是拒绝这个 promise；但 resolve(..) 既可能完成 promise，也可能拒绝，要根据传入参数而定。如果传给 resolve(..) 的是一个非 Promise、非 thenable 的立即值，这个 promise 就会用这个值完成。

但是，如果传给 resolve(..) 的是一个真正的 Promise 或 thenable 值，这个值就会被递归展开，并且（要构造的）promise 将取用其最终决议值或状态。

2. Promise.resolve(..) 和 Promise.reject(..)

创建一个已被拒绝的 Promise 的快捷方式是使用 Promise.reject(..)，Promise.resolve(..) 常用于创建一个已完成的 Promise。

3. then(..) 和 catch(..)

每个 Promise 实例（不是 Promise API 命名空间）都有 then(..) 和 catch(..) 方法，通过这两个方法可以为这个 Promise 注册完成和拒绝处理函数。Promise 决议之后，立即会调用这两个处理函数之一，但不会两个都调用，而且总是异步调用。

then(..) 接受一个或两个参数：第一个用于完成回调，第二个用于拒绝回调。catch(..) 只接受一个拒绝回调作为参数，并自动替换默认完成
回调。then(..) 和 catch(..) 也会创建并返回一个新的 promise，这个 promise 可以用于实现Promise 链式流程控制。

4. Promise.all([ .. ]) 和 Promise.race([ .. ])

对 Promise.all([ .. ]) 来说，只有传入的所有 promise 都完成，返回 promise 才能完成。如果有任何 promise 被拒绝，返回的主 promise 就立即会被拒绝。

对 Promise.race([ .. ]) 来说，只有第一个决议的 promise（完成或拒绝）取胜，并且其决议结果成为返回 promise 的决议。

当心！若向 Promise.all([ .. ]) 传入空数组，它会立即完成，但 Promise.race([ .. ]) 会挂住，且永远不会决议。

## 小结
Promise 非常好，请使用。它们解决了我们因只用回调的代码而备受困扰的控制反转问题。

它们并没有摈弃回调，只是把回调的安排转交给了一个位于我们和其他工具之间的可信任的中介机制。

Promise 链也开始提供（尽管并不完美）以顺序的方式表达异步流的一个更好的方法，这有助于我们的大脑更好地计划和维护异步 JavaScript 代码。我们将在第 4 章看到针对这个问题的一种更好的解决方案！
